### ISSUE TITLE
git ext_pillar dev branches, how to use for development?

### ISSUE THREAD

**Issue by Paulo-Nunes** [saltstack/salt#51564]
### Description of Issue/Question
I have state files in one git repo and pillar data in another git repo.
Within the pillar repo I have 3 different environments each with their own top.sls. They are separate and do not interact with each other.
I would like to be able to make use of pillarenv on the cli to test dev branches for pillar data.

pillar repo (master branch)
A/top.sls
B/top.sls
C/top.sls

I want to create a branch from master and edit only contents of C. let us call it Cdev.
then on the cli I want to run `pillar.items pillarenv=Cdev` and see the pillar in Cdev . Equally important is to run `pillar.items` on a minion that normally gets pillar data from C and have that data be exclusively from C and not Cdev.

When I run the command, though, I get what appears to be the merging of the 2 branches.
How can I configure salt to do what I want?

### Setup
(Please provide relevant configs and/or SLS files (Be sure to remove sensitive info).)
Sample top.sls for the C directory on master:
```
C:
  '*':
    - apache
```
I tried modifying this in the Cdev branch to:
```
Cdev:
  '*':
    - apache
```

```
ext_pillar:
  - git:
    - master git@github.com:salt/salt_A.git:
      - name: A
      - env: A
      - root: A
    - master git@github.com:salt/salt_B.git:
      - name: B
      - env: B
      - root: B
    - master git@github.com:salt/salt_C.git:
      - name: C
      - env: C
      - root: C
    - Cdev git@github.com:salt/salt_A.git:
      - name: Cdev
      - env: Cdev
      - root: C
```
Ideally I would not want to have to add a new item to the list above for each branch I want to test.
```
Master conf:
	file_roots:
		test:
			- /srv/salt
	top_file_merging_strategy: same
	default_top: base
	fileserver_backend:
		- git
		- roots
	gitfs_provider: pygit2
	pillar_source_merging_strategy: (tried with smart and none)
	pillarenv_from_saltenv: False
	git_pillar_provider: pygit2
```	
Minion has defaults but:
	I tried setting pillar env, but I don't want this behavior.

### Steps to Reproduce Issue
(Include debug logs if possible and relevant.)
Create the pillar structure like above and add a pillar to Cdev that is unique to Cdev.

`salt "minion" pillar.items`
	shows all the pillar data, but the pillar unique to Cdev is also included. What I want is, if no pillarenv is specified to execute as if it is the master branch. With merging set to none I expect it to not merge all the pillars, but it did anyway.
`salt "minion" pillar.items pillarenv=C`
	Shows only what is in C, as expected
`salt "minion" pillar.items pillarenv=Cdev`
	Shows what is in Cdev as expected

### Versions Report
(Provided by running `salt --versions-report`. Please also mention any differences in master/minion versions.)
```
Salt Version:
           Salt: 2018.3.2
 
Dependency Versions:
           cffi: Not Installed
       cherrypy: Not Installed
       dateutil: Not Installed
      docker-py: Not Installed
          gitdb: Not Installed
      gitpython: Not Installed
          ioflo: Not Installed
         Jinja2: 2.8.1
        libgit2: 0.20.0
        libnacl: Not Installed
       M2Crypto: Not Installed
           Mako: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.4.6
   mysql-python: 1.2.5
      pycparser: Not Installed
       pycrypto: 2.6.1
   pycryptodome: Not Installed
         pygit2: 0.20.3
         Python: 2.7.14 (default, Jan 31 2018, 02:12:13)
   python-gnupg: Not Installed
         PyYAML: 3.11
          PyZMQ: 14.5.0
           RAET: Not Installed
          smmap: Not Installed
        timelib: Not Installed
        Tornado: 4.2.1
            ZMQ: 4.0.5
 
System Versions:
           dist: redhat 6.10 Santiago
         locale: UTF-8
        machine: x86_64
        release: 2.6.32-754.9.1.el6.x86_64
         system: Linux
        version: Red Hat Enterprise Linux Server 6.10 Santiago
```

**Comment by Ch3LL** [comment]
ping @terminalmage any ideas here?

**Comment by terminalmage** [comment]
`top_file_merging_strategy` has nothing to do with pillar, it only applies to state top files.

This is just how pillar works. All environments are merged into a single dictionary. If you want a specific pillar environment you must specify the `pillarenv` argument. Note that you can only really do this when running `pillar.items` or when running states.

**Comment by Paulo-Nunes** [comment]
I was more concerned with `pillar_source_merging_strategy`. I included `top_file_merging_strategy` for completeness.

So you are saying that there is no way to have a dev branch for my pillar data if I am already using `env` in the ext_pillar configuration?  If I set `pillarenv` on the minion config, that will be the default env unless specified on the CLI?
Last I checked, managing minion configs and restarting them from a state was not straightforward. Has this changed?  In order to use the `pillarenv` config on the minion I would have to have a state that applies that property to the minion and restarts it before executing the rest of the states. I have 3 disjoint groups of minions that read from A,B,C respectively.

There is no way to have similar behavior to states where specifying `saltenv` will allow you to run that branch of code?
